### PR TITLE
feat: Add filter_policy to pgvector integration

### PR DESCRIPTION
**Pull request by vblagoje**
### Why:
Adds flexible filtering options to pgvector integration. By introducing a filter policy option (`replace` or `merge`), developers can now control how runtime filters are applied relative to initialization time filters. 

- partially fixes https://github.com/deepset-ai/haystack-core-integrations/issues/780
- supersedes https://github.com/deepset-ai/haystack-core-integrations/pull/781

### What:
- Added `filter_policy` parameter with options `replace` and `merge` across multiple retrievers to control filter behavior dynamically.
- Unit tests were updated or added to validate the new functionality.

### How can it be used:

- **Dynamic Filter Behavior Adjustment:**  Users can decide whether to completely override the initial filters set during the retriever's initialization (`replace`) or merge them with runtime filters, with the latter taking precedence (`merge`).

- **Complex Search Scenarios:**  
    - In cases where the context of a query might dictate altering pre-set filters without discarding them, the `merge` option allows for an additive approach.
    - For strict query contexts that require ignoring initial filters, the `replace` option offers a clean slate for filters at runtime.

### How did you test it:
- Unit tests were enhanced or newly created to cover both `replace` and `merge` scenarios for the `filter_policy` parameter. 
- Tests ensure that filter logic is correctly applied based on the policy setting, whether it merges runtime filters with initial filters or replaces them entirely.
- Additional test cases should be considered for complex filter merge scenarios to ensure priority and override mechanisms function as expected.

### CODE REVIEW

File: integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/embedding_retriever.py [hunk 1]
```diff
@@ -75,7 +76,9 @@ def __init__(
             and the most similar documents are the ones with the smallest score.
             **Important**: if the document store is using the `"hnsw"` search strategy, the vector function
             should match the one utilized during index creation to take advantage of the index.
-
+        :param filter_policy: Policy to determine how filters are applied. Defaults to "replace".
```
Comment (anakin87) [inline integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/embedding_retriever.py:79]:
```suggestion
        :param filter_policy: Policy to determine how filters are applied.
```

Lately, we have stopped mentioning defaults in docstrings, if they can be easily inferred from the signature.

File: integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/embedding_retriever.py [hunk 2]
```diff
@@ -139,7 +144,11 @@ def run(
 
         :returns: List of Documents similar to `query_embedding`.
         """
-        filters = filters or self.filters
+        if self.filter_policy == "merge" and filters:
```
Comment (anakin87) [inline integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/embedding_retriever.py:147]: In the docstrings of the `run` method, I would mention that the way runtime filters are applied depends on the `filter_policy` chosen at document store initialization.

**Comment by vblagoje** [comment]
Good feedback, thanks @anakin87 , will apply across the board

**Review by anakin87 — COMMENTED** [review]
Could you also add a test as requested/done in https://github.com/deepset-ai/haystack-core-integrations/pull/823#discussion_r1646184590?

**Comment by vblagoje** [comment]
> Could you also add a test as requested/done in [#823 (comment)](https://github.com/deepset-ai/haystack-core-integrations/pull/823#discussion_r1646184590)?

Yes, already on it - for all PRs 😓

**Comment by anakin87** [comment]
> > Could you also add a test as requested/done in [#823 (comment)](https://github.com/deepset-ai/haystack-core-integrations/pull/823#discussion_r1646184590)?
> 
> Yes, already on it - for all PRs 😓

I understand your pain.
Perhaps you could create a utility function to handle filters in the Haystack repository and test it only there once and for all.

**Review by anakin87 — APPROVED** [review]

**Comment by vblagoje** [comment]
@anakin87 we need to use the already released https://github.com/deepset-ai/haystack/blob/main/haystack/document_stores/types/filter_policy.py instead of literal, across the board as well. I'll add that change here.

**Comment by vblagoje** [comment]
Please delay review until a new patch release of 2.2.x is released. We need these [two PRs](https://github.com/deepset-ai/haystack/commits/main/haystack/document_stores/types/filter_policy.py) to be included in a patch release first.

**Comment by vblagoje** [comment]
Ready for review now @anakin87 

- Since you last looked at this PR we introduced `apply_filter_policy` function that is used by all Retrievers in this [PR](https://github.com/deepset-ai/haystack/pull/7902) and we thoroughly tested that function in that PR. We then just applied it here in run method
- Please note if FilterPolicy type is used consistently both in init and in run, and that it is documented well

**Comment by vblagoje** [comment]
Stefano recommended one more small improvement. Please hold your review until a new commit with it is added here

File: integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/embedding_retriever.py [hunk 3]
```diff
@@ -104,6 +110,7 @@ def to_dict(self) -> Dict[str, Any]:
             filters=self.filters,
             top_k=self.top_k,
             vector_function=self.vector_function,
+            filter_policy=self.filter_policy.value if self.filter_policy else None,
```
Comment (anakin87) [inline integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/embedding_retriever.py:113]:
```suggestion
            filter_policy=self.filter_policy.value,
```

File: integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/keyword_retriever.py [hunk 4]
```diff
@@ -78,6 +83,7 @@ def to_dict(self) -> Dict[str, Any]:
             self,
             filters=self.filters,
             top_k=self.top_k,
+            filter_policy=self.filter_policy.value if self.filter_policy else None,
```
Comment (anakin87) [inline integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/keyword_retriever.py:86]:
```suggestion
            filter_policy=self.filter_policy.value,
```

File: integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/keyword_retriever.py [hunk 5]
```diff
@@ -106,13 +114,19 @@ def run(
...
         :param top_k: Maximum number of Documents to return.
 
         :returns: A dictionary with the following keys:
             - `documents`: List of `Document`s that match the query.
         """
-        filters = filters or self.filters
+        if self.filter_policy == FilterPolicy.MERGE and filters:
+            filters = {**self.filters, **filters}
+        else:
+            filters = filters or self.filters
```
Comment (anakin87) [inline integrations/pgvector/src/haystack_integrations/components/retrievers/pgvector/keyword_retriever.py:128]:
```suggestion
        filters = apply_filter_policy(self.filter_policy, self.filters, filters)
```

**Review by anakin87 — COMMENTED** [review]
please incorporate the suggestions and merge :slightly_smiling_face:

**Comment by vblagoje** [comment]
@anakin87 one last look 🙏

**Review by anakin87 — COMMENTED** [review]
looks good

**Pull request merged**